plenum/server/replica.py

```diff
@@ -355,13 +364,66 @@ def primaryName(self, value: Optional[str]) -> None:
                 self.removeObsoletePpReqs()
             self._stateChanged()
 
-    def primaryChanged(self, primaryName, lastOrderedPPSeqNo):
-        if self.lastOrderedPPSeqNo < lastOrderedPPSeqNo:
-            self.lastOrderedPPSeqNo = lastOrderedPPSeqNo
+    def primaryChanged(self, primaryName):
         self.primaryName = primaryName
-        if primaryName == self.name:
-            assert self.lastOrderedPPSeqNo >= lastOrderedPPSeqNo
-            self._lastPrePrepareSeqNo = self.lastOrderedPPSeqNo
+        self._lastPrePrepareSeqNo = 0
+        self.set_last_ordered_for_non_master()
+
+    def shouldParticipate(self, viewNo: int, ppSeqNo: int) -> bool:
+        """
+        Replica should only participating in the consensus process and the
+        replica did not stash any of this request's 3-phase request
+        """
+        return self.node.isParticipating and (viewNo, ppSeqNo) \
+                                             not in self.stashingWhileCatchingUp
+
+    def on_view_change_start(self):
+        assert self.isMaster
+        lst = self.last_prepared_certificate_in_view(self.viewNo)
+        self.last_prepared_before_view_change = lst
+        logger.debug('{} setting last prepared for master to {}'.format(self, lst))
+
+    def on_view_change_done(self):
+        assert self.isMaster
+        self.last_prepared_before_view_change = None
+
+    def get_lowest_probable_prepared_certificate_in_view(self, view_no) -> Optional[int]:
+        """
+        Return lowest pp_seq_no of the view for which can be prepared but
+        choose from unprocessed PRE-PREPAREs and PREPAREs.
+        """
+        # TODO: Naive implementation, dont need to iterate over the complete
+        # data structures, fix this later
+        seq_no_pp = SortedList()      # pp_seq_no of PRE-PREPAREs
+        # pp_seq_no of PREPAREs with count of PREPAREs for each
+        seq_no_p = set()
+
+        for (v, p) in self.prePreparesPendingPrevPP:
+            if v == view_no:
+                seq_no_pp.add(p)
+            if v > view_no:
+                break
+
+        for (v, p), pr in self.preparesWaitingForPrePrepare.items():
+            if v == view_no and len(pr) >= 2*self.f:
+                seq_no_p.add(p)
+
+        for n in seq_no_pp:
+            if n in seq_no_p:
+                return n
+        return None
+
+    def set_last_ordered_for_non_master(self):
+        if not self.isMaster:
+            # If not master instance choose last ordered seq no to be 1 less
+            # the lowest prepared certificate in this view
+            lowest_prepared = self.get_lowest_probable_prepared_certificate_in_view(
```

Comment:
`If not master instance choose last ordered seq no to be 1 less the lowest prepared certificate in this view`

Why?